sycl/source/handler.cpp

```diff
@@ -108,8 +108,13 @@ event handler::finalize() {
   if (getCGTypeVersion(MCGType) >
       static_cast<unsigned int>(detail::CG::CG_VERSION::V0)) {
     // If there were uses of set_specialization_constant build the kernel_bundle
+    bool HasSpecConstArgs =
+        std::any_of(MArgs.begin(), MArgs.end(), [](detail::ArgDesc Arg) {
+          return Arg.MType == detail::kernel_param_kind_t::
+                                  kind_specialization_constants_buffer;
+        });
     std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImpPtr =
-        getOrInsertHandlerKernelBundle(/*Insert=*/false);
+        getOrInsertHandlerKernelBundle(/*Insert=*/HasSpecConstArgs);
```

Comment:
Suggest using type trait https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/cg_types.hpp#L160 in handler.hpp to detect if a lambda takes kernel handler at create kernel bundle there. It should save us some run time.

Reply:
Sadly I don't believe the type information of the kernel lambda is available in `finalize()` at compile time. I could save it from `StoreLambda`, but that would require a new member in the `handle` class, which in turn would break `abi/layout_handler.cpp`. Did you have something else in mind?

Reply:
Can we do the check and call ` getOrInsertHandlerKernelBundle(/*Insert=*/true);` in the `StoreLambda`?